### PR TITLE
Support for multiple data saving backends in liveplotting (video mode)

### PR DESCRIPTION
**Pull request by damiencrielaard**
Hi @sldesnoo-Delft!

We discussed adding support for different datasaving backends in the video mode code, so I drafted something please take a look! 

Some interesting things:

- The way I have it now, I add a quantify_core dependency to core_tools. The only way I see to avoid this is to instead build infrastructure allow to define these backends externally and pass these to the liveplotting.
- I find that I need the `UnravelMultiParameter` to convert the internal multiparameter description of the scan to convert it to something that plays nice with Quantify.

Please let me know how this aligns with how you envisioned this.

------------------------------

## Description of PR:

Adds the ability to specify a data saving backend for the liveplotting and implements a quantify backend.

### Explanation of changes

- Adds a `data_saving_backend` keyword argument to `liveplotting`. 
- Moves the data saving code to a `qcodes` module.
- Adds data saving using `quantify_core`.

### Motivation of changes

We wish to save in the format of a quantify dataset, whereas the liveplotting uses qcodes. This flexibility is desirable.

### CODE REVIEW

**Comment by sldesnoo-Delft** [comment]
Hi @damiencrielaard,
Can we have a meeting on the changes?

Adding a quantify dependency is a no-go. We are not gonna use quantify and I don't want unnecessary dependencies in the code, especially not a package that is not yet stable. There are already too much dependencies and too much unused code is loaded during startup. (And honestly, I still believe quantify is a poor derivative of qcodes.)

There are 3 solutions for the data_saver configuration: 
a) pass data_saver as a string and conditionally import the specific data saver when liveplotting is started or when save is pressed. This is easy for the user.
b) pass a data_saver method to the liveplotting initializer. This is more flexible. It's better engineering, but makes an ugly interface.
c) globally set the data saver class for the core_tools GUI. 
  from core_tools.storage_backends.quantify import DataSaver
  data_saver = DataSaver(path='./data')
  core_tools.GUI.set_data_saver(data_saver)
I have a preference for option c.

Note: there are already 2 different backends: core_tools database and qcodes dataset. Quantify will be the third. So the backend options will be 'core_tools', 'qcodes', 'quantify'.

The AwgScanToQuantifyMapper should be in its own file. Actually, I wish it wasn't needed at all. So it should be easy to remove when quantify has been fixed.

**Comment by damiencrielaard** [comment]
Hi @sldesnoo-Delft 

Thanks for taking a look. I think your concerns are valid, so indeed let's discuss how I can best address them. 

I've sent you an email to plan a meeting.

File: core_tools/GUI/keysight_videomaps/data_saver/qcodes.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+        Args:
+            vm_data_parameter: a MultiParameter instance describing the measurement with settables, gettables and setpoints.
+            label: a string that is used to label the dataset.
+
+        Returns:
+            A Tuple (ds, metadata) containing the created dataset ds and a metadata dict with information about the dataset.
+        """
+        measure = Measure(vm_data_parameter)
+        data = measure.run(quiet=True, name=label, location=self.path)
```
Comment (sldesnoo-Delft) [inline core_tools/GUI/keysight_videomaps/data_saver/qcodes.py:32]: location must be unique per measurement. It's better not to specify it and let it be generated automatically by qcodes.

File: core_tools/GUI/keysight_videomaps/data_saver/native.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        Performs a measurement using core tools and writes the data to disk.
+
+        Args:
+            vm_data_parameter: a MultiParameter instance describing the measurement with settables, gettables and setpoints.
+            label: a string that is used to label the dataset.
+
+        Returns:
+            A Tuple (ds, metadata) containing the created dataset ds and a metadata dict with information about the dataset.
+        """
+        assert hasattr(sample_info, 'project')
```
[the remark this answers is not in the document]
Comment (sldesnoo-Delft) [reply core_tools/GUI/keysight_videomaps/data_saver/native.py:38]: Actually, in this case it is better just to try to read sample_info.project. When the database connection is not properly configured this already throws a ConnectionError.

**Review by sldesnoo-Delft — COMMENTED** [review]
I've added a few remarks on the code. Most of them can be solved by just removing some code.

File: core_tools/GUI/keysight_videomaps/data_saver/data_saver.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+from pathlib import Path
+
+from qcodes import MultiParameter
+
+
+class IDataSaver(ABC):
+    """Specifies the interface data savers are to adhere to."""
+
+    @abstractmethod
+    def __init__(self, path: Optional[Union[Path, str]] = None):
```
[the remark this answers is not in the document]
Comment (damiencrielaard) [reply core_tools/GUI/keysight_videomaps/data_saver/data_saver.py]: I removed it here. That also makes it unnecessary in the concrete classes. So I just removed the `__init__` from all.

File: core_tools/GUI/keysight_videomaps/data_saver/native.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (damiencrielaard) [reply core_tools/GUI/keysight_videomaps/data_saver/native.py]: I changed it to reading all the `sample_info` and add them to the metadata. That way we are relying on the ConnectionError being raised like you said, plus it gets added to the notes in the ppt which may be useful.

File: core_tools/GUI/keysight_videomaps/data_saver/data_saver.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    def save_data(self, vm_data_parameter: MultiParameter, label: str) -> Tuple[Any, Dict[str, Any]]:
+        """
+        Saves the data to disk.
+
+        Args:
+            vm_data_parameter: a MultiParameter instance describing the measurement with settables, gettables and setpoints.
+            label: a string that is used to label the dataset.
+
+        Returns:
+            A Tuple (ds, metadata) containing the created dataset ds and a metadata dict with information about the dataset.
```
[the remark this answers is not in the document]
Comment (damiencrielaard) [reply core_tools/GUI/keysight_videomaps/data_saver/data_saver.py]: I changed the return type to `Dict[str,str]`

File: core_tools/GUI/keysight_videomaps/data_saver/qcodes.py (hunk 1 above)
Comment (damiencrielaard) [reply core_tools/GUI/keysight_videomaps/data_saver/qcodes.py]: Done

**Comment by damiencrielaard** [comment]
Hi @sldesnoo-Delft!

Thanks a lot for the review 😄 . I think I addressed your comments. Please take a look when you can.

File: core_tools/data/SQL/connect.py [hunk 5]
```diff
@@ -24,6 +24,9 @@ def __init__(self, project, set_up, sample):
         sample_info.set_up = set_up
         sample_info.sample = sample
 
+    def __str__(self) -> str:
```
Comment (sldesnoo-Delft) [inline core_tools/data/SQL/connect.py:27]: Nice addition :-)

**Review by sldesnoo-Delft — APPROVED** [review]
Looks good.

**Pull request merged**